File: 8.string-to-integer-atoi.py
#
# @lc app=leetcode id=8 lang=python3
#
# [8] String to Integer (atoi)
#

# Time:  O(n)
# Space: O(1)

# @lc code=start
class Solution:
    def myAtoi(self, s: str) -> int:
        i, res = 0, 0
        while i < len(s) and s[i] == " ":
            i += 1
        if i == len(s):
            return res

        sign = 1
        if s[i] == "+":
            i += 1
        elif s[i] == "-":
            sign = -1
            i += 1

        INT_MAX, INT_MIN = 2147483647, -2147483648
        while i < len(s) and "0" <= s[i] <= "9":
            if res > (INT_MAX - int(s[i])) // 10:
                return INT_MAX if sign > 0 else INT_MIN
            res = res * 10 + int(s[i])
            i += 1
        return sign * res


# @lc code=end

File: 8.string-to-integer-atoi.js
/*
 * @lc app=leetcode id=8 lang=javascript
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function (s) {
  let i = 0;
  while (i < s.length && s[i] === " ") {
    ++i;
  }
  if (i === s.length) {
    return 0;
  }

  let sign = s[i] === "-" ? -1 : 1;
  if (s[i] === "+" || s[i] === "-") {
    ++i;
  }

  let res = 0;
  while (i < s.length && s[i].match(/\d/)) {
    if (res > (2147483647 - parseInt(s[i])) / 10) {
      return sign === 1 ? 2147483647 : -2147483648;
    }
    res = res * 10 + parseInt(s[i++]);
  }
  return sign * res;
};
// @lc code=end
